Source/AMRInterpolator/SphericalExtraction.hpp

```diff
@@ -0,0 +1,139 @@
+/* GRChombo
+ * Copyright 2012 The GRChombo collaboration.
+ * Please refer to LICENSE in GRChombo's root directory.
+ */
+
+#ifndef SPHERICALEXTRACTION_HPP_
+#define SPHERICALEXTRACTION_HPP_
+
+#include "SphericalGeometry.hpp"
+#include "SphericalHarmonics.hpp"
+#include "SurfaceExtraction.hpp"
+
+//! A child class of SurfaceExtraction for extraction on spherical shells
+class SphericalExtraction : public SurfaceExtraction<SphericalGeometry>
+{
+  public:
+    struct params_t : SurfaceExtraction::params_t
+    {
+        int &num_extraction_radii = num_surfaces;
+        std::vector<double> &extraction_radii = surface_param_values;
+        int &num_points_theta = num_points_u;
+        int &num_points_phi = num_points_v;
+        std::array<double, CH_SPACEDIM> center; //!< the center of the spherical
+                                                //!< shells
+        std::array<double, CH_SPACEDIM> &extraction_center = center;
+        int num_modes; //!< the number of modes to extract
+        std::vector<std::pair<int, int>> modes; //!< the modes to extract
+                                                //!< l = first, m = second
+        // constructor
+        params_t() = default;
+
+        // copy constructor defined due to references pointing to the wrong
+        // things with the default copy constructor
+        params_t(const params_t &params)
+            : center(params.center), num_modes(params.num_modes),
+              modes(params.modes), SurfaceExtraction::params_t(params)
+        {
+        }
+    };
+    const std::array<double, CH_SPACEDIM> m_center;
+    const int m_num_modes;
+    const std::vector<std::pair<int, int>> m_modes;
+
+    SphericalExtraction(const params_t &a_params, double a_dt, double a_time,
+                        bool a_first_step, double a_restart_time = 0.0)
+        : SurfaceExtraction(a_params.center, a_params, a_dt, a_time,
+                            a_first_step, a_restart_time),
+          m_center(a_params.center), m_num_modes(a_params.num_modes),
+          m_modes(a_params.modes)
+    {
+    }
+
+    SphericalExtraction(const params_t &a_params,
+                        const std::vector<std::pair<int, Derivative>> &a_vars,
+                        double a_dt, double a_time, bool a_first_step,
+                        double a_restart_time = 0.0)
+        : SphericalExtraction(a_params, a_dt, a_time, a_first_step,
+                              a_restart_time)
+    {
+        add_vars(a_vars);
+    }
+
+    SphericalExtraction(const params_t &a_params,
+                        const std::vector<int> &a_vars, double a_dt,
+                        double a_time, bool a_first_step,
+                        double a_restart_time = 0.0)
+        : SphericalExtraction(a_params, a_dt, a_time, a_first_step,
+                              a_restart_time)
+    {
+        add_vars(a_vars);
+    }
```

Comment:
It it possible to have a constructor like this without the `a_first_step` (set to `a_dt==a_time` as is done for Weyl)? Just because I was testing it out, building my own example to test your code, and I had to go to the Weyl Extraction constructor to remember how this was done...

Reply:
I'm not going to do this because there are already quite a few constructors in this class and I don't want to confuse it more and also if you are writing your own child class of this, then you will probably copy and modify `WeylExtraction` anyway.